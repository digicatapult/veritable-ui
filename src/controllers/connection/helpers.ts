import { ILogger } from '../../logger.js'
import { ConnectionRow } from '../../models/db/types.js'

export const checkDb = async (rows: ConnectionRow[], initialPinAttemptsRemaining: number, logger: ILogger) => {
  const [connectionCheck] = rows
  if (initialPinAttemptsRemaining > 0 && connectionCheck.pin_tries_remaining_count == 0) {
    logger.debug('Maximum number of pin attempts has been reached.')
    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message:
        'Maximum number of pin attempts has been reached, please reach out to the company you are attempting to connect to.',
      nextScreen: 'error',
    }
  } else if (connectionCheck.status === 'verified_us') {
    logger.debug('Pin has been verified.')

    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message: 'Success',
      nextScreen: 'success',
    }
  } else if (
    connectionCheck.pin_tries_remaining_count < initialPinAttemptsRemaining ||
    (initialPinAttemptsRemaining === 0 && connectionCheck.pin_tries_remaining_count === 4)
  ) {
    logger.debug(`Pin was invalid remaining attempts number:${connectionCheck.pin_tries_remaining_count}.`)
    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message: `Sorry, your code is invalid. You have ${connectionCheck.pin_tries_remaining_count} attempts left before the PIN expires.`,
      nextScreen: 'form',
    }
  }
  logger.debug(`Pollig: No change in db detected.`)

  return false
}
