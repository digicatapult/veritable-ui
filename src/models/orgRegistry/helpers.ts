import z from 'zod'
import { InternalError } from '../../errors.js'
import { ILogger } from '../../logger.js'
import { RegistryType } from '../db/types.js'
import { OrganisationProfile, OrganisationRequest, Registries, SharedOrganisationInfo } from './organisationRegistry.js'
import { companyHouseResultSchema } from './registrySchemas/companyHouseSchema.js'
import { nyStateResultSchema } from './registrySchemas/NYStateSchema.js'
import { openCorporatesResultSchema } from './registrySchemas/openCorporatesSchema.js'

export const makeOrganisationRegistryRequest = async (
  route: string,
  registryName: string,
  apiKey: string | null
): Promise<unknown> => {
  const url = new URL(route)

  const headers: HeadersInit = {}
  if (apiKey !== null) {
    headers.Authorization = apiKey
  }

  const response = await fetch(url.toString(), {
    method: 'GET',
    headers,
  })
  const contentType = response.headers.get('content-type')
  if (response.ok) {
    if (contentType && contentType.includes('application/json')) {
      return response.json()
    }
    throw new InternalError(`Unexpected content-type ${contentType} from ${registryName} API, for ${url.toString()}`)
  }

  if (response.status === 404) {
    return null
  }

  throw new InternalError(`Error calling ${registryName} API`)
}
const retrieveEndpointAndSchema = (
  orgReq: OrganisationRequest,
  registry: Registries[RegistryType]
): { url: string; schema: z.ZodSchema; apiKey?: string } => {
  switch (orgReq.selectedRegistry) {
    case 'company_house':
      return {
        url: `${registry.url}/company/${encodeURIComponent(orgReq.companyNumber)}`,
        schema: companyHouseResultSchema,
        apiKey: registry.api_key,
      }

    case 'ny_state':
      return { url: `${registry.url}?dos_id=${orgReq.companyNumber}`, schema: nyStateResultSchema }

    case 'open_corporates':
      return {
        url: `${registry.url}/companies/${orgReq.registryCountryCode.toLowerCase()}/${encodeURIComponent(orgReq.companyNumber)}?api_token=${registry.api_key}`,
        schema: openCorporatesResultSchema,
      }

    default:
      throw new InternalError(`Registry ${orgReq.selectedRegistry} not set`)
  }
}

export const organisationRegistryProfile = async (
  orgReq: OrganisationRequest,
  registry: Registries[RegistryType],
  logger: ILogger
): Promise<OrganisationProfile> => {
  logger.info(`Getting ${registry.registry_name} profile for ${orgReq.companyNumber}`)
  const urlAndSchema = retrieveEndpointAndSchema(orgReq, registry)
  const companyProfile = await makeOrganisationRegistryRequest(
    urlAndSchema.url,
    registry.registry_name,
    urlAndSchema.apiKey ? urlAndSchema.apiKey : null
  )
  if (companyProfile === null) {
    return { type: 'notFound' }
  }
  try {
    const result = urlAndSchema.schema.parse(companyProfile)
    const sharedInfo = result as SharedOrganisationInfo
    return { type: 'found', company: sharedInfo }
  } catch (error) {
    logger.error(`Error parsing ${registry.registry_name} profile for ${orgReq.companyNumber}: ${error}`)
    return { type: 'notFound' }
  }
}
