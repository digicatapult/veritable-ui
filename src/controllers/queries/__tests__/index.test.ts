import { expect } from 'chai'
import { describe, it } from 'mocha'
import sinon from 'sinon'

import { QueriesController } from '../index.js'
import { mockIds, toHTMLString, withQueriesMocks } from './helpers.js'

describe('QueriesController', () => {
  afterEach(() => {
    sinon.restore()
  })
  describe('queries', () => {
    it('should match the snapshot of the rendered query page', async () => {
      const { args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const result = await controller.queries().then(toHTMLString)
      expect(result).to.equal('queries_template')
    })
    it('should call db as expected', async () => {
      const { dbMock, args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.queryManagement().then(toHTMLString)
      expect(spy.firstCall.calledWith('connection', [], [['updated_at', 'desc']])).to.equal(true)
      expect(spy.secondCall.calledWith('query', {}, [['updated_at', 'desc']])).to.equal(true)
    })
    it('should call db as expected', async () => {
      const { dbMock, args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.queryManagement('VER123').then(toHTMLString)
      const search = 'VER123'
      const query = search ? [['company_name', 'ILIKE', `%${search}%`]] : {}
      expect(spy.firstCall.calledWith('connection', query, [['updated_at', 'desc']])).to.equal(true)

      expect(spy.secondCall.calledWith('query', {}, [['updated_at', 'desc']])).to.equal(true)
    })

    it('should call db as expected', async () => {
      const { dbMock, args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.scope3CarbonConsumption().then(toHTMLString)
      expect(spy.firstCall.calledWith('connection', [], [['updated_at', 'desc']])).to.equal(true)
    })

    it('should call db as expected', async () => {
      const { dbMock, args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.scope3CarbonConsumption('VER123').then(toHTMLString)
      const search = 'VER123'
      const query = search ? [['company_name', 'ILIKE', `%${search}%`]] : {}
      expect(spy.firstCall.calledWith('connection', query, [['updated_at', 'desc']])).to.equal(true)
    })

    it('should call page with stage FORM as expected', async () => {
      const { args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionStage({ connectionId: 'connection-id', action: 'form' })
        .then(toHTMLString)

      expect(result).to.equal('scope3_form_scope3')
    })

    it('should call page with stage success as expected', async () => {
      const { args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionStage({
          connectionId: 'connection-id',
          action: 'success',
          productId: 'SomeID',
          quantity: 111,
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_success_scope3')
    })

    it('should call page with stage error if rpc fails', async () => {
      const { args, cloudagentMock } = withQueriesMocks()
      cloudagentMock.submitDrpcRequest = sinon.stub().rejects(new Error())

      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionStage({
          connectionId: 'connection-id',
          action: 'success',
          productId: 'SomeID',
          quantity: 111,
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_error_scope3')
    })

    it('should call page with stage error if rpc succeeds without response', async () => {
      const { args, cloudagentMock } = withQueriesMocks()
      cloudagentMock.submitDrpcRequest = sinon.stub().resolves(undefined)

      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionStage({
          connectionId: 'connection-id',
          action: 'success',
          productId: 'SomeID',
          quantity: 111,
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_error_scope3')
    })

    it('should call page with stage error if rpc succeeds with error', async () => {
      const { args, cloudagentMock } = withQueriesMocks()
      cloudagentMock.submitDrpcRequest = sinon.stub().resolves({
        error: new Error('error'),
        id: 'request-id',
      })

      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionStage({
          connectionId: 'connection-id',
          action: 'success',
          productId: 'SomeID',
          quantity: 111,
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_error_scope3')
    })
  })
  describe('query responses', () => {
    it('should call db as expected', async () => {
      const { args, dbMock } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.scope3CarbonConsumptionResponse('SomeId').then(toHTMLString)
      const search = 'SomeId'
      expect(spy.firstCall.calledWith('query', { id: search })).to.equal(true)
    })

    it('should call query response page with stage FORM as expected', async () => {
      const { args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const result = await controller.scope3CarbonConsumptionResponse('someId123').then(toHTMLString)

      expect(result).to.equal('queriesResponse_template')
    })
    it('should call query response page with stage SUCCESS as expected', async () => {
      const { args } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionResponseSubmit('5390af91-c551-4d74-b394-d8ae0805059e', {
          companyId: '2345789',
          action: 'success',
          totalScope3CarbonEmissions: '25',
        })
        .then(toHTMLString)

      expect(result).to.equal('queriesResponse_template')
    })
    it('should call page with stage error if rpc succeeds without response', async () => {
      const { args, cloudagentMock } = withQueriesMocks()
      cloudagentMock.submitDrpcRequest = sinon.stub().resolves(undefined)

      const controller = new QueriesController(...args)

      const result = await controller
        .scope3CarbonConsumptionResponseSubmit('5390af91-c551-4d74-b394-d8ae0805059e', {
          companyId: '2345789',
          action: 'success',
          totalScope3CarbonEmissions: '25',
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_error_scope3')
    })
    it('should call page with stage error if rpc fails', async () => {
      const { args, cloudagentMock } = withQueriesMocks()
      cloudagentMock.submitDrpcRequest = sinon.stub().rejects(new Error())

      const controller = new QueriesController(...args)
      const result = await controller
        .scope3CarbonConsumptionResponseSubmit('5390af91-c551-4d74-b394-d8ae0805059e', {
          companyId: '2345789',
          action: 'success',
          totalScope3CarbonEmissions: '25',
        })
        .then(toHTMLString)

      expect(result).to.equal('scope3_error_scope3')
    })
  })
  it('should call page with stage error if rpc succeeds with error', async () => {
    const { args, cloudagentMock } = withQueriesMocks()
    cloudagentMock.submitDrpcRequest = sinon.stub().resolves({
      error: new Error('error'),
      id: 'request-id',
    })

    const controller = new QueriesController(...args)
    const result = await controller
      .scope3CarbonConsumptionResponseSubmit('5390af91-c551-4d74-b394-d8ae0805059e', {
        companyId: '2345789',
        action: 'success',
        totalScope3CarbonEmissions: '25',
      })
      .then(toHTMLString)

    expect(result).to.equal('scope3_error_scope3')
  })
  describe('viewing query responses', () => {
    it('should call db as expected', async () => {
      const { args, dbMock } = withQueriesMocks()
      const controller = new QueriesController(...args)
      const spy = dbMock.get
      await controller.scope3CarbonConsumptionViewResponse('SomeId').then(toHTMLString)
      const search = 'SomeId'
      expect(spy.firstCall.calledWith('query', { id: search })).to.equal(true)
    })
  })
  describe('Partial Query', () => {
    const { args, dbMock } = withQueriesMocks()
    let result: string

    before(async () => {
      sinon.restore()

      const controller = new QueriesController(...args)
      result = await controller
        .scope3CO2Partial(
          mockIds.queryId, // url param
          mockIds.companyId, // url param
          'on' // partialSelect query string param
        )
        .then(toHTMLString)
    })

    afterEach(() => sinon.restore())

    describe('if query string param [partialQuery] not provided', () => {
      it('renders a regular form template', async () => {
        const controller = new QueriesController(...args)
        result = await controller
          .scope3CO2Partial(
            mockIds.queryId, // url param
            mockIds.companyId // url param
          )
          .then(toHTMLString)

        expect(result).to.be.equal('queriesResponse_template')
      })
    })

    it('returns the correct HTML template', () => {
      expect(result).to.contain('queriesResponse_template')
    })

    it('retrieves connections and query details from a database', async () => {
      expect(dbMock.get.args).to.deep.equal([
        ['query', { id: '00000000-0000-0000-0000-d8ae0805059e' }],
        ['connection', { id: 'cccccccc-0001-0000-0000-d8ae0805059e' }],
        ['connection', { status: 'verified_both' }],
      ])
    })

    it('pulls connections and returns along with partial = true', async () => {
      const formatted = JSON.parse(result.replace('queriesResponse_template-', ''))

      expect(result).to.contain('queriesResponse_template-')
      expect(formatted.partial).to.be.equal(true)
      expect(formatted.connections).to.deep.equal([
        {
          agent_connection_id: 'aaaaaaaa-0000-0000-0000-d8ae0805059e',
          company_name: 'PARTIAL_QUERY',
          id: 'cccccccc-0000-0000-0000-d8ae0805059e',
          status: 'verified_both',
        },
        {
          agent_connection_id: 'aaaaaaaa-0000-0000-0000-d8ae0805059e',
          company_name: 'VERIFIED_THEM',
          id: 'cccccccc-0000-0000-0000-d8ae0805059e',
          status: 'verified_them',
        },
      ])
    })
  })
})
