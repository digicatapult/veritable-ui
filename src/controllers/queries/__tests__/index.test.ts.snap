// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueriesController queries should match the snapshot of the rendered query page 1`] = `
Readable {
  "_events": Object {
    "close": undefined,
    "data": undefined,
    "end": undefined,
    "error": undefined,
    "readable": undefined,
  },
  "_maxListeners": undefined,
  "_read": [Function],
  "_readableState": ReadableState {
    "awaitDrainWriters": null,
    "buffer": Array [],
    "bufferIndex": 0,
    "highWaterMark": 16,
    "length": 0,
    "pipes": Array [],
    Symbol(kState): 1052941,
  },
  Symbol(shapeMode): true,
  Symbol(kCapture): false,
}
`;
